vowpalwabbit/core/src/reductions/cb/cb_explore_adf_graph_feedback.cc

```diff
@@ -0,0 +1,527 @@
+      g.ones();
+
+      if ((z / _gamma) < 0.f) { g[_fhat.size()] = -1.f; }
+    }
+  }
+};
+
+bool valid_graph(const std::vector<VW::cb_graph_feedback::triplet>& triplets)
+{
+  // return false if all triplet vals are zero
+  for (auto& triplet : triplets)
+  {
+    if (triplet.val != 0.f) { return true; }
+  }
+  return false;
+}
+
+std::pair<arma::mat, arma::vec> set_initial_coordinates(const arma::vec& fhat, float gamma)
+{
+  // find fhat min
+  auto min_fhat = fhat.min();
+  arma::vec gammafhat = gamma * (fhat - min_fhat);
+
+  // initial p can be uniform random
+  arma::mat coordinates(gammafhat.size() + 1, 1);
+  for (size_t i = 0; i < gammafhat.size(); i++) { coordinates[i] = 1.f / gammafhat.size(); }
+
+  // initial z can be 1
+  // but also be nice if all fhat's are zero
+  float z = gamma * (1 - (min_fhat == 0 ? 1.f / fhat.size() : min_fhat));
+
+  coordinates[gammafhat.size()] = z;
+
+  return {coordinates, gammafhat};
+}
+
+arma::vec get_probs_from_coordinates(arma::mat& coordinates, const arma::vec& fhat, VW::workspace& all)
+{
+  // constraints are enforcers but they can be broken, so we need to check that probs are positive and sum to 1
+
+  // we also have to check for nan's because some starting points combined with some gammas might make the constraint
+  // optimization go off the charts; it should be rare but we need to guard against it
+
+  size_t num_actions = coordinates.n_rows - 1;
+  auto count_zeros = 0;
+  bool there_is_a_one = false;
+  bool there_is_a_nan = false;
+
+  for (size_t i = 0; i < num_actions; i++)
+  {
+    if (VW::math::are_same(static_cast<float>(coordinates[i]), 0.f) || coordinates[i] < 0.f)
+    {
+      coordinates[i] = 0.f;
+      count_zeros++;
+    }
+    if (coordinates[i] > 1.f)
+    {
+      coordinates[i] = 1.f;
+      there_is_a_one = true;
+    }
+    if (std::isnan(coordinates[i])) { there_is_a_nan = true; }
+  }
+
+  if (there_is_a_nan)
+  {
+    for (size_t i = 0; i < num_actions; i++) { coordinates[i] = 1.f - fhat(i); }
+  }
+
+  if (there_is_a_one)
+  {
+    for (size_t i = 0; i < num_actions; i++)
+    {
+      if (coordinates[i] != 1.f) { coordinates[i] = 0.f; }
+    }
+  }
+
+  float p_sum = 0;
+  for (size_t i = 0; i < num_actions; i++) { p_sum += coordinates[i]; }
+
+  if (!VW::math::are_same(p_sum, 1.f))
+  {
+    float rest = 1.f - p_sum;
+    float rest_each = rest / (num_actions - count_zeros);
+    for (size_t i = 0; i < num_actions; i++)
+    {
+      if (coordinates[i] == 0.f) { continue; }
+      else { coordinates[i] = coordinates[i] + rest_each; }
+    }
+  }
+
+  float sum = 0;
+  arma::vec probs(num_actions);
+  for (size_t i = 0; i < probs.n_rows; ++i)
+  {
+    probs(i) = coordinates[i];
+    sum += probs(i);
+  }
+
+  if (!VW::math::are_same(sum, 1.f))
+  {
+    // leaving this here just in case this happens for some reason that we did not think to check for
+    all.logger.err_warn("Probabilities do not sum to 1, they sum to: " + std::to_string(sum));
```

Comment:
New code should just use `warn`

```suggestion
    all.logger.warn("Probabilities do not sum to 1, they sum to: {}", sum);
```